### ISSUE TITLE
Ability to limit parallelism for tasks that require shared resources

### ISSUE THREAD

**Issue by shevek** [gradle/gradle#7047]
### Expected Behavior

Parallel builds should succeed even when tasks require a large amount of resources.

### Current Behavior

gradle --parallel crashes with OOME: "Java Heap Space" because tasks are clashing on access to a shared resource. In our case, it's system RAM - we have a task which requires a lot of system RAM, we have about 20 such tasks in the project, so we can't run them all in parallel. We would like to annotate this task definition somehow so that no more than one instance of it runs at a time, OR, possibly, that it never runs in parallel with ANY other task.

* The task cannot be rewritten to be more efficient - if that were possible, it would be a major coup in pure mathematics.
* We cannot add more RAM - we have 256Gb of RAM. We need to restrict max cardinality.
* We don't really want to reduce overall parallelism - we lose a lot of time to javac and other tasks, and we have 64 CPUs which we'd like to use.

### Context

It's causing our builds to take 30 minutes (fully sequential) instead of about 6 (by amdahl's law)

See also https://discuss.gradle.org/t/disabling-parallel-for-certain-tasks-only/10151 which appears to be non-conclusive.

Can we please have an annotation like CacheableTask for NonParallelizableTask?

**Comment by shevek** [comment]
Unreliable workaround:
```
./gradlew expensiveTask # runs in each submodule
./gradlew --parallel assemble
```

This relies on:
* expensiveTask using `@CacheableTask`, which it does
* that nothing else interfering with that caching, which is *not* guaranteed.
* that most of the parallelizable work runs *after, not before* expensiveTask. Not generally true, but true for us.

**Comment by oehme** [comment]
For your specific use case you could try using the [Worker API](https://guides.gradle.org/using-the-worker-api/) with Process isolation and an explicit heap requirement. This way the expensive task is separated from the rest of the work and Gradle has explicit knowledge about its requirement and so it will not start that work if that goes beyond the available system memory.

Let me know if that helps.

**Comment by shevek** [comment]
We currently have 35 instances of ExpensiveTask in our task-graph.

I suppose the other hack would be to simply hold a global static mutex in the task and just block gradle.

**Comment by shevek** [comment]
Question: How does gradle ensure that multiple parallel test tasks don't run, or is that also open season? As in, if our tests require access to an external resource, we can't build parallel anyway?

**Comment by shevek** [comment]
That sounds reasonable.

I suspect that in the case of CPU, the only realistic values are "one" and "all". 1 is single-threaded code. "all" is correct multi-threaded code. Any value in between implies inefficiently written multi-threaded code. A weaker, but similar argument might be made for RAM.

**Comment by shevek** [comment]
Holding a mutex actually doesn't solve the problem because if we run on an 8-CPU system, we run 8 threads, but gradle runs 7 javac threads as well, and this causes (again) the OS scheduler and cache and so forth to have a worse than 8x degradation. We really _do_ need to be able to tell Gradle to keep its hands in its pockets for a little while.

**Comment by oehme** [comment]
You can use a java.util.concurrent.Semaphore with max-workers permits as I sketched above. You'd add a `doFirst`/`doLast` to every task to acquire/release one of those permits. Your special task would require all permits at once.

**Comment by oehme** [comment]
Of course this should be in the core, I was just trying to give you a workaround. Adding this workaround to all tasks in a build is easy:

```
allprojects {
  tasks.configureEach {
    doFirst { acquirePermit }
    doLast { releasePermit }
  }
}
```

**Comment by atorok** [comment]
I would like to +1. Having the worker API and making it easier to use is great, but it doesn't immediately help with all the existing tasks that would need to be converted in third party  plugins and existing tools that are ran part of the build that already know how to fully utilize the CPU. 

DSL to deal with this like being proposed here would be a great addition.
 
A simple way  to look at it conceptually is to say how many workers should be idling when a task runs. I guess it depends on how the build is structured, for me this seems conceptually simpler (and Gradle can implement the leases behind the scenes), but there might be use-cases I can't think of that wouldn't be properly addressed by this. 

@oehme  is that something you would consider a PR for ? If so what's the proper way to have a discussion to make sure it ends up being something that can be merged.

**Comment by atorok** [comment]
Unfortunately the workaround is not feasible for us as the resource intensive "task" is actually a series of tasks and it's not always clear which one is last. We also rely on cleanup by a `finalizedBy` task, which with `--parallel` is delayed to the very end causing resources not to free up when run like that.

**Comment by oehme** [comment]
I'm afraid a PR would be pretty hard, as scheduling is one of the hairiest parts of the code base. 

I didn't understand why the workaround wouldn't work for you. You make all your expensive tasks require `n` leases, thus making sure nothing runs in parallel to them. Every one acquires and releases those leases using doFirst and doLast as shown above. You don't keep leases across several tasks.

**Comment by atorok** [comment]
It doesn't work because we do keep leases across several tasks and the order of these is dynamic, so it's hard to follow which is last. If we knew which one is last we could avoid using `finalizedBy` for clean-up. We can avoid most of the issues by restructuring the build into fewer tasks to achieve a more coarse granularity, which we started to do, but it's a lot of work. 

Anyway the point I was trying to make is that it would be nice if the solution for this problem would not expose low level primitives like aquire/release but something closer to the Gradle model ( e.x. idling workers, mutually exclusive tasks or task types, etc )

**Comment by shevek** [comment]
Another task which uses all CPUs is proguard - if you're building multiple dists (rare, but we do it), you can't run proguard in gradle parallel.

**Comment by oehme** [comment]
> It doesn't work because we do keep leases across several tasks and the order of these is dynamic, so it's hard to follow which is last

Why are you keeping leases across several tasks? Acquire/release them before/after every task instead and it would work.

**Comment by shevek** [comment]
I could see a situation where an external mutable resource needed to be owned by a set of tasks, but I would agree that's bad practice and not what Gradle is intended for - should be a single task.

Although some situations with testing make me think ...

**Comment by atorok** [comment]
Thanks for following this @oehme !

> Why are you keeping leases across several tasks? Acquire/release them before/after every task instead and it would work.

A simplified setup looks like this: 
The "leases" are elasticsearch clusters and the tasks run the same tests, identically configured `Test` tasks.  The purpose is to assert that the tests pass despite of changes to the cluster that happen in between (like upgrading or partially upgrading a cluster).

We have many such tasks, and there's nothing special about the last test , nor the first one.
So this could only be solved by being specific when we configure them, but that's easy to miss and error prone. It would be nice to have something that we could apply in a centralized manner across the build. 

This setup has been working well but it poses some challenges with `--parallel` besides the resource constraints ( sorry don't mean to derail this issue, just including for complete context ). We rely on `finalizedBy` to shut down the clusters after the last test, but with `--parallel` that seems to happen too late and we end up with tens of ES nodes running out of memory even on well equipped workstations. 
To solve this  we came up with some DSL to allow tasks to declare the clusters they use, then do a ref count and use pre and post task hooks to bring them up and down automatically. The custom plugin is here https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java and this test has some usage examples https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/testKit/testclusters/build.gradle . 
I guess the same could be reused to implement the semaphores as long as the pre and post task hooks are executed by the same worker as the tasks themselves, I haven't checked how this is done. 

@shevek I think it's more a function of what the task does, even beyond testing, one might want to split a longer task into two smaller ones if the inputs can be split to get some benefits from up-to-date checks. e.x Changing input1 requires a rebuild, but changing input2 requires rebuilding only 90% . If input 1 doesn't change often enough having two tasks makes sense, especially if "loading" the shared resource carries a cost. This is essentially what my tests do, they take the shared resource trough a sequence of states A - B -> C -> D where each state requires going trough the previous one so breaking things up  differently would cause a lot of additional work.

**Comment by atorok** [comment]
> Holding a mutex actually doesn't solve the problem because if we run on an 8-CPU system, we run 8 threads, but gradle runs 7 javac threads as well, and this causes (again) the OS scheduler and cache and so forth to have a worse than 8x degradation. We really _do_ need to be able to tell Gradle to keep its hands in its pockets for a little while.

Having hooks like "worker starting" and "finishing"  would allow us to implement this.
I think that would play nicely with how we are able to hook into various other parts of Gradle.
@oehme do you think that makes sense / something easier to do ?

**Comment by shevek** [comment]
This is currently killing us, and making it very hard to construct reliable Gradle builds. We can run sequential, which takes 10x the time to build, or run a weird hack which is unreliable and still loses 50% efficiency, i.e. 2 hours of productive dev time per build. Resources we are currently battling are:

* Simple single-threaded tasks, e.g. javac, jar, etc.
* Tasks which download slowly from network and use no system resource at all, e.g. building license lists.
* Tasks which use a lot of threads, e.g. parser generators, individual multi-threaded tests.
* Tasks which use a lot of memory, e.g. spotbugs can't parallelize well in 64Gb RAM
* Tasks which need external mutable resources, e.g. database-backed tests

It's basically impossible to construct an efficient gradle build file for any project which contains a mixture of such tasks.

Please consider additionally the desire for an individual junit test to be able to say that it's multithreaded, whereas another junit test may say that it is not.

What is the chance of any movement on this issue, please?

**Comment by shevek** [comment]
Another example: errorprone: With 24Gb of RAM, I can run most tasks, even regular javac, with a parallelism of 12, but errorprone uses more RAM per thread, so errorprone tasks can only run with a parallelism of 7. They need to declare their RAM usage, and block instantiation of additional tasks until they release the RAM. Otherwise, the entire build is limited to a parallelism of 7.

**Comment by oehme** [comment]
Fixed by https://github.com/gradle/gradle/pull/9914

**Comment by oehme** [comment]
@ghale I just noticed that #9914 didn't include documentation or release notes - Can you add something before the release?

**Comment by ghale** [comment]
@oehme That was intentional.  We've discussed it further and decided to make some more changes (both to the api as well as the internals) before making this public.  Once we've made those changes, we'll add docs, samples, etc.

@shevek Stay tuned, we are actively working on this stuff and expect to have something you can try out soon.

**Comment by big-guy** [comment]
@adammurdoch is there something in 6.1 related to this?

**Comment by adammurdoch** [comment]
Only a refactored API for the stuff in #9914.

**Comment by oehme** [comment]
This issue can be closed, since we have the[ `BuildService`](https://docs.gradle.org/current/javadoc/org/gradle/api/services/BuildService.html) API now.
